finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/TlsClientPrep.scala

```diff
@@ -0,0 +1,168 @@
+package com.twitter.finagle.buoyant
+
+import com.twitter.finagle._
+import com.twitter.finagle.client.Transporter
+import com.twitter.finagle.ssl.{Engine, Ssl}
+import com.twitter.finagle.transport.{TlsConfig, Transport}
+import java.io.FileInputStream
+import java.net.{InetSocketAddress, SocketAddress}
+import java.security.KeyStore
+import java.security.cert.{CertificateFactory, X509Certificate}
+import javax.net.ssl.{SSLContext, TrustManagerFactory}
+import scala.util.Random
+
+object TlsClientPrep {
+
+  object role extends Stack.Role("TlsClientPrep") {
+    val finagle = Stack.Role("TlsClientPrep.Finagle")
+  }
+
+  val description = "Configures per-endpoint TLS settings"
+
+  type Params = Stack.Params
+
+  case class TlsEnabled(enabled: Boolean)
+  implicit object TlsEnabled extends Stack.Param[TlsEnabled] {
+    val default = TlsEnabled(false)
+  }
+
+  sealed trait Trust
+  implicit object Trust extends Stack.Param[Trust] {
+    case class Verified(name: String, certs: Seq[X509Certificate]) extends Trust
+    object UnsafeNotVerified extends Trust
+    object NotConfigured extends Trust
+
+    val default = NotConfigured
+  }
+
+  /**
+   * May be extended to implement a TlsClientPrep module. Supports
+   * Params-driven TLS configuration.
+   */
+  type Module[Req, Rsp] = Stack.Module[ServiceFactory[Req, Rsp]]
+  type Stk[Req, Rsp] = Stack[ServiceFactory[Req, Rsp]]
+  type Stkable[Req, Rsp] = Stackable[ServiceFactory[Req, Rsp]]
+
+  trait TlsTrustModule[Req, Rsp] extends Stack.Module[ServiceFactory[Req, Rsp]] {
+    override val role = TlsClientPrep.role
+    override val description = TlsClientPrep.description
+    override val parameters = Nil
+
+    def tlsEnabled: TlsEnabled
+    def trust: Trust
+
+    def make(params: Stack.Params, next: Stk[Req, Rsp]): Stk[Req, Rsp] =
+      Stack.Leaf(role, next.make(params + tlsEnabled + trust))
+  }
+
+  /** A module that uses the given `commonName` for all requests. */
+  def static[Req, Rsp](cn: String, caCert: Option[String]): Stkable[Req, Rsp] =
+    new TlsTrustModule[Req, Rsp] {
+      override val tlsEnabled = TlsEnabled(true)
+      override val trust = Trust.Verified(cn, caCert.toSeq.map(loadCert(_)))
+    }
+
+  /** A module that configures TLS to ignore certificate validation. */
+  def withoutCertificateValidation[Req, Rsp]: Stkable[Req, Rsp] =
+    new TlsTrustModule[Req, Rsp] {
+      override val tlsEnabled = TlsEnabled(true)
+      override val trust = Trust.UnsafeNotVerified
+    }
+
+  /** A module that always disables TLS. */
+  def disable[Req, Rsp]: Stkable[Req, Rsp] =
+    new TlsTrustModule[Req, Rsp] {
+      override val tlsEnabled = TlsEnabled(false)
+      override val trust = Trust.NotConfigured
+    }
+
+  def finagleDisabled[Req, Rsp]: Stkable[Req, Rsp] =
```

Comment:
Consider moving this below finagleModule and add a comment indicating that this is an alternative to finagleModule that disables the finagle tls params.